solanaj/src/main/java/org/p2p/solanaj/kits/transaction/parser/SerumSwapInstructionParser.kt

```diff
@@ -0,0 +1,132 @@
+package org.p2p.solanaj.kits.transaction.parser
+
+import org.p2p.solanaj.kits.transaction.SwapDetails
+import org.p2p.solanaj.kits.transaction.network.ConfirmedTransactionRootResponse
+import org.p2p.solanaj.kits.transaction.network.meta.InstructionResponse
+import org.p2p.solanaj.programs.SerumSwapProgram
+import java.lang.Exception
+import java.math.BigInteger
+
+class SerumSwapInstructionParser {
+
+    class SerumSwapParseError(message: String) : Exception(message)
+
+    fun isTransactionContainsSerumSwap(transactionRoot: ConfirmedTransactionRootResponse): Boolean {
+        val instructions = transactionRoot.transaction?.message?.instructions.orEmpty()
+        return getSerumSwapInstructionIndex(instructions) != -1
+    }
+
+    fun parse(
+        signature: String,
+        transactionRoot: ConfirmedTransactionRootResponse,
+    ): Result<SwapDetails> {
+        val parsedTransaction = transactionRoot.transaction
+            ?: return parseError("($signature) .transaction is null")
+        val instructions = parsedTransaction.message.instructions
+        val swapInstructionIndex = getSerumSwapInstructionIndex(instructions)
+
+        val preTokenBalances = transactionRoot.meta.preTokenBalances
+
+        // get swapInstruction
+        val swapInstruction = instructions.getOrNull(swapInstructionIndex)
+            ?: return parseError("($signature) swapInstruction is null")
+
+        // get all mints
+        val mints = preTokenBalances.map { it.mint }.distinct().toMutableList()
+        if (mints.size < 2) return parseError("($signature) preTokenBalance.mint < 2")
+
+        // transitive swap: remove usdc or usdt if exists
+        if (mints.size == 3) {
+            mints.removeAll(this::isMintUsdx)
+        }
+
+        // define swap type
+        val isTransitiveSwap = mints.none(this::isMintUsdx)
+
+        // assert
+        val accounts = swapInstruction.accounts
+        if (accounts.isEmpty()) return parseError("($signature) swapInstruction.accounts == 0")
+
+        if (isTransitiveSwap && accounts.size != 27) {
```

Comment:
do we know why it is checked on 27?
any way i suggest to place 27 / 16 to consts also

Reply:
I don't understand any of this code really :D 
Maybe @1sakov can tell us smth